mfa/mfatokenprovider.go

```diff
@@ -0,0 +1,36 @@
+package mfa
+
+import (
+	"fmt"
+)
+
+// TokenProvider is an interface to provide an mfa token. It's intended that provided do whatever is necessary to get
+// a token, eg prompt the use via the terminal or fetch it from a yubikey.
+type TokenProvider interface {
+	Retrieve(mfaSerial string) (string, error)
+}
```

Comment:
Perhaps `type TokenProvider func() (string, error)` would be simpler (and would match the AWS SDK)

Reply:
At some level the serial is needed as any kind of tool that can generate mfa tokens for multiple things needs the serial to know which token to generate. In `SessionTokenProvider` and `AssumeRoleProvider` the mfa token is useless on its own as you also have to pass the mfa serial when authing. So the serial has be stored somewhere, I've considered expanding the interface to include the serial but haven't come up with anything I like (suggestions welcome). Possibly clunky but the obvious choice is something along these lines, could  simplify code elsewhere though.

```
type TokenProvider interface {
    GetToken() string, error
    GetSerial() string, error
    SetSerial(serial string)
}
```

Where do you see that signature in the aws SDK, can't find it, would like to see that usage?

Reply:
I've refactored `mfa.TokenProvider` to include serial, it has simplified other code too.